lib/srv/alpnproxy/forward_proxy.go

```diff
@@ -0,0 +1,395 @@
+	Listener net.Listener
+	// CloseContext is the close context.
+	CloseContext context.Context
+	// Handlers is a list of CONNECT request handlers.
+	Handlers []ConnectRequestHandler
+}
+
+// CheckAndSetDefaults checks and sets default config values.
+func (c *ForwardProxyConfig) CheckAndSetDefaults() error {
+	if c.Listener == nil {
+		return trace.BadParameter("missing listener")
+	}
+	if c.CloseContext == nil {
+		return trace.BadParameter("missing close context")
+	}
+	return nil
+}
+
+// ForwardProxy is a forward proxy that serves CONNECT tunnel requests.
+type ForwardProxy struct {
+	cfg ForwardProxyConfig
+}
+
+// NewForwardProxy creates a new forward proxy server.
+func NewForwardProxy(cfg ForwardProxyConfig) (*ForwardProxy, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &ForwardProxy{
+		cfg: cfg,
+	}, nil
+}
+
+// Start starts serving on the listener.
+func (p *ForwardProxy) Start() error {
+	err := http.Serve(p.cfg.Listener, p)
+	if err != nil && !utils.IsUseOfClosedNetworkError(err) {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// Close closes the forward proxy.
+func (p *ForwardProxy) Close() error {
+	if err := p.cfg.Listener.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// GetAddr returns the listener address.
+func (p *ForwardProxy) GetAddr() string {
+	return p.cfg.Listener.Addr().String()
+}
+
+// ServeHTTP serves HTTP requests. Implements http.Handler.
+func (p *ForwardProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	// Only allow CONNECT tunnel requests. Reject if clients send plain HTTP.
+	if !IsConnectRequest(req) {
+		rw.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	clientConn := hijackClientConnection(rw)
+	if clientConn == nil {
+		log.Error("Failed to hijack client connection.")
+		rw.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	defer clientConn.Close()
+
+	for _, handler := range p.cfg.Handlers {
+		if handler.Match(req) {
+			handler.Handle(p.cfg.CloseContext, clientConn, req)
+			return
+		}
+	}
+
+	writeHeaderToHijackedConnection(clientConn, req.Proto, http.StatusBadRequest)
+}
+
+// ForwardToHostHandlerConfig is the config for ForwardToHostHandler.
+type ForwardToHostHandlerConfig struct {
+	// Match returns true if this handler wants to handle the provided request.
+	MatchFunc func(req *http.Request) bool
+
+	// Host is the destination to forward the request to. If empty, the request
+	// is forwarded to its original host.
+	Host string
+}
+
+// SetDefaults sets default config values.
+func (c *ForwardToHostHandlerConfig) SetDefaults() {
+	if c.MatchFunc == nil {
+		c.MatchFunc = MatchAllRequests
+	}
+}
+
+// MatchAllRequests is a MatchFunc that returns true for all requests.
+func MatchAllRequests(req *http.Request) bool {
+	return true
+}
+
+// MatchAWSRequests is a MatchFunc that returns true if request is an AWS API
+// request.
+func MatchAWSRequests(req *http.Request) bool {
```

Comment:
Seems to be unused.

Reply:
good catch. suppose to use it in `tsh/aws.go`. will update